client/sftp_client.py

```diff
@@ -56,100 +93,429 @@ def signal_sftp_start(self, _):
+		self.treeview_transfer.connect('row-expanded', self.signal_tv_expand_row)
+
+		self.treeview_transfer.set_model(self._tv_model)
+		self.treeview_transfer.show_all()
 
 		self.popup_menu = Gtk.Menu.new()
+
+		self.menu_item_paused = Gtk.CheckMenuItem.new_with_label('Paused')
+		menu_item = self.menu_item_paused
+		menu_item.connect('toggled', self.signal_menu_toggled_paused)
+		self.popup_menu.append(menu_item)
+
+		self.menu_item_cancel = Gtk.MenuItem.new_with_label('Cancel')
+		menu_item = self.menu_item_cancel
+		menu_item.connect('activate', self.signal_menu_activate_cancel)
+		self.popup_menu.append(menu_item)
+
+		menu_item = Gtk.SeparatorMenuItem()
+		self.popup_menu.append(menu_item)
+
 		menu_item = Gtk.MenuItem.new_with_label('Clear')
-		menu_item.connect('activate', self.log_clear)
+		menu_item.connect('activate', self.signal_menu_activate_clear)
 		self.popup_menu.append(menu_item)
 		self.popup_menu.show_all()
 
-	def _signal_treeview_button_pressed(self, _, event):
+	def _get_selected_tasks(self):
+		treepaths = self._get_selected_treepaths()
+		if treepaths is None:
+			return None
+		# Find the task selected task then find tall the tasks in the queue with it as a parent.
+		parent_task = [task for task in self.queue.queue if task.treerowref is not None and task.treerowref.valid() and task.treerowref.get_path() in treepaths][0]
+		children = [task for task in self.queue.queue if task.parents is not None and parent_task in task.parents]
+		parent_task = [parent_task,]
+		return parent_task + children
+
+	def _get_selected_treepaths(self):
+		selection = self.treeview_transfer.get_selection()
+		model, treeiter = selection.get_selected()
+		if treeiter is None:
+			return None
+		treepaths = []
+		treepaths.append(model.get_path(treeiter))
+		return treepaths
+
+	def _change_task_state(self, state_from, state_to):
+		with self.queue.mutex:
+			for task in self._get_selected_tasks():
+				if task.state in state_from:
+					if state_to == 'Cancelled':
+						if isinstance(task, TransferDirectoryTask):
+							if task.has_children:
+								# Folder clicked indicates a parent task has been clicked and to freeze its progress
+								task.folder_clicked = True
+						elif task.parents is not None:
+							for parent in task.parents:
+								if not parent.folder_clicked:
+									parent.size -= 1
+					task.state = state_to
+
+	def _sync_view(self):
+		# This value was set to True to prevent the treeview from freezing.
+		if not self.queue.mutex.acquire(blocking=True):
+			return
+		if not self._tv_lock.acquire(blocking=False):
+			self.queue.mutex.release()
+			return
+		for task in self.queue.queue:
+			# Make task visible if it has no parents (it is a root task) or if its parent has been expanded
+			if task.parents is None:
+				task.show = True
+			else:
+				parent_task = task.parents[len(task.parents)-1]
+				task.show = bool(parent_task.show_children)
+			if not task.show:
```

Comment:
Why not just add the task to the model and allow GTK to handle whether or not it should be shown based on if the user has expanded the parent node(s).